src/zope/container/ordered.py

```diff
@@ -272,17 +273,14 @@ def updateOrder(self, order):
         if len(order) != len(self._order):
             raise ValueError("Incompatible key set.")
 
-        was_dict = {}
-        will_be_dict = {}
-        new_order = PersistentList()
+        order = [checkAndConvertName(x) for x in order]
 
-        for i, obj in enumerate(order):
-            was_dict[self._order[i]] = 1
-            will_be_dict[obj] = 1
-            new_order.append(obj)
-
-        if will_be_dict != was_dict:
+        if frozenset(order) != frozenset(self._order):
             raise ValueError("Incompatible key set.")
 
-        self._order = new_order
+        if order == self._order:
+            return
+
+        self._order.sort(key=order.index)
```

Comment:
It's `O(n log n)` from what I understand about timsort, not quadratic, but I see your point about it being clever.

I did that because I really wanted to be able to write the test line:
```
        # It also kept the exact key objects as given
        self.assertIs(oc.keys()[0], keys[1])
```

But I suppose there are enough other checks that I can feel safe without that particular pair of suspenders.